File: lib/tools/firecrawl.ts
import { getCached, setCache } from '../supabase';
import type { ToolResult, ScrapedPage } from './types';
import { buildToolResult } from './fallback';
import { assessScrapeQuality } from './scrape-quality';
import { getPolicyForDomain, computeRetryDelay } from './retry-policy';
import { isPlaywrightScrapeEnabled, scrapeWithPlaywright } from './playwright-scrape';

const BASE_URL = 'https://api.firecrawl.dev/v1';

// Extraction prompt templates by page type
const EXTRACT_PROMPTS: Record<string, string> = {
  pricing: 'Extract pricing tiers, features per tier, costs, and billing periods. Highlight any discounts or special offers.',
  features: 'Extract product features, capabilities, technical details, and key differentiators.',
  competitor: 'Extract positioning, USPs, target audience, and competitive claims.',
  review: 'Extract review ratings, sentiment, pros/cons, and buyer feedback.',
  generic: 'Extract key product information, pricing, features, and any competitive claims.',
};

function delay(ms: number): Promise<void> {
  return new Promise<void>((resolve) => {
    setTimeout(() => resolve(), ms);
  });
}

/**
 * Infer extraction profile from URL to use targeted prompts.
 */
function selectExtractPrompt(url: string): string {
  const lower = url.toLowerCase();
  if (/pricing|plan|billing|cost/.test(lower)) return EXTRACT_PROMPTS.pricing;
  if (/feature|capability|docs|technical/.test(lower)) return EXTRACT_PROMPTS.features;
  if (/review|g2|capterra|comparison/.test(lower)) return EXTRACT_PROMPTS.review;
  if (/competitor|vs|alternative/.test(lower)) return EXTRACT_PROMPTS.competitor;
  return EXTRACT_PROMPTS.generic;
}

/**
 * Call Firecrawl with standard options.
 */
async function firecrwlFetch(url: string, extractPrompt: string, apiKey: string, isStrict = false): Promise<{ markdown: string; title: string } | null> {
  try {
    const res = await fetch(`${BASE_URL}/scrape`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify({
        url,
        formats: ['markdown', 'extract'],
        extract: { prompt: extractPrompt },
        ...(isStrict && { waitForSelector: 'body', onlyMainContent: true, removeTags: ['script', 'style'] }),
      }),
    });

    if (!res.ok) return null;

    const raw = await res.json() as any;
    const markdown: string = raw.data?.markdown ?? '';
    const title: string = raw.data?.metadata?.title ?? url;

    return markdown.trim().length > 50 ? { markdown, title } : null;
  } catch {
    return null;
  }
}

export async function scrapePage(url: string): Promise<ToolResult<ScrapedPage>> {
  try {
    const { assertSafeUrl } = await import('@/lib/guardrails');
    await assertSafeUrl(url);
  } catch {
    return buildToolResult({
      data: { url, title: '', markdown: '', excerpt: '' },
      status: 'failed',
      source: 'url-policy',
      sourceUrl: url,
      operation: 'scrapePage',
    });
  }

  const cacheKey = `scrape:${url}`;
  const cached = await getCached('firecrawl', cacheKey);
  if (cached) {
    return { ...(cached as ToolResult<ScrapedPage>), cached: true };
  }

  const apiKey = process.env.FIRECRAWL_API_KEY;
  if (!apiKey) {
    // Fallback: basic fetch + strip HTML
    return scrapeBasic(url);
  }

  const policy = getPolicyForDomain(url);
  const extractPrompt = selectExtractPrompt(url);
  let markdown = '';
  let title = url;
  let attemptsMade = 0;

  // ── Multi-attempt strategy with escalating fallbacks ─────────────────────
  // Attempt 1: Standard Firecrawl
  let result = await firecrwlFetch(url, extractPrompt, apiKey, false);
  attemptsMade++;

  if (!result && policy.useFirecrawlStrict && attemptsMade < policy.maxAttempts) {
    // Attempt 2: Strict Firecrawl (for flaky sites like LinkedIn)
    await delay(computeRetryDelay(policy, attemptsMade));
    result = await firecrwlFetch(url, extractPrompt, apiKey, true);
    attemptsMade++;
  }

  let usedPlaywright = false;
  if (
    !result &&
    policy.useBrowserFallback &&
    isPlaywrightScrapeEnabled() &&
    attemptsMade < policy.maxAttempts
  ) {
    await delay(computeRetryDelay(policy, attemptsMade));
    const pw = await scrapeWithPlaywright(url);
    attemptsMade++;
    if (pw) {
      result = { markdown: pw.markdown, title: pw.title };
      usedPlaywright = true;
    }
  }

  if (!result && attemptsMade < policy.maxAttempts) {
    // Scrape.do (free tier — anti-bot, proxy rotation, optional JS render)
    await delay(computeRetryDelay(policy, attemptsMade));
    result = await scrapeDoFetch(url);
    attemptsMade++;
  }

  if (!result && attemptsMade < policy.maxAttempts) {
    // Smart direct fetch (rotating UA, realistic headers, structured extraction)
    await delay(computeRetryDelay(policy, attemptsMade));
    result = await smartDirectFetch(url);
    attemptsMade++;
  }

  if (result) {
    markdown = result.markdown;
    title = result.title;
  }

  // ── Quality assessment ──────────────────────────────────────────────────
  const quality = assessScrapeQuality(markdown, url);

  // ── Build result with quality metadata ──────────────────────────────────
  const page: ScrapedPage = {
    url,
    title,
    markdown,
    excerpt: markdown.slice(0, 500),
  };

  // Determine status based on quality + attempts
  let status: 'ok' | 'degraded' | 'failed' = 'ok';
  let confidence = 0.85;
  let sourceLabel = 'Firecrawl';

  if (!markdown || markdown.trim().length === 0) {
    status = 'failed';
    confidence = 0.15;
  } else if (quality.isBlockPage) {
    status = 'degraded';
    confidence = 0.35;
  } else if (!quality.isValid) {
    status = 'degraded';
    confidence = quality.qualityScore * 0.7; // cap degraded at ~0.7
  } else if (usedPlaywright) {
    status = 'degraded';
    confidence = 0.75;
    sourceLabel = 'Playwright';
  } else if (attemptsMade > 1) {
    status = 'degraded';
    confidence = 0.7; // fallback used but content ok
  }

  const result_final = buildToolResult<ScrapedPage>({
    data: page,
    status,
    source: sourceLabel,
    sourceUrl: url,
    confidenceOverride: confidence,
  });

  await setCache('firecrawl', cacheKey, result_final);
  return result_final;
}

// ── Scrape.do fallback (free tier: 1000 req/month, anti-bot + proxy rotation) ──
const SCRAPE_DO_BASE = 'https://api.scrape.do';

function needsJsRender(url: string): boolean {
  const spaPatterns = [
    /facebook\.com/i, /linkedin\.com/i, /twitter\.com/i, /x\.com/i,
    /app\./i, /dashboard\./i, /portal\./i,
  ];
  return spaPatterns.some(p => p.test(url));
}

async function scrapeDoFetch(url: string): Promise<{ markdown: string; title: string } | null> {
  try {
    const { assertSafeUrl } = await import('@/lib/guardrails');
    await assertSafeUrl(url);
  } catch {
    return null;
  }

  const token = process.env.SCRAPE_DO_TOKEN;
  if (!token) return null;

  try {
    const params = new URLSearchParams({
      token,
      url,
      output: 'markdown',
    });
    if (needsJsRender(url)) {
      params.set('render', 'true');
      params.set('waitUntil', 'domcontentloaded');
    }

    const res = await fetch(`${SCRAPE_DO_BASE}/?${params.toString()}`, {
      signal: AbortSignal.timeout(25_000),
    });

    if (!res.ok) return null;

    const markdown = await res.text();
    if (!markdown || markdown.trim().length < 50) return null;

    const titleMatch = markdown.match(/^#\s+(.+)$/m);
    const title = titleMatch?.[1]?.trim() || url;

    return { markdown: markdown.trim().slice(0, 8000), title };
  } catch {
    return null;
  }
}

// ── Smart direct fetch with rotating UA + structured extraction ──────────────
const USER_AGENTS = [
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36',
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 14_5) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.5 Safari/605.1.15',
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:127.0) Gecko/20100101 Firefox/127.0',
  'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36',
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 14_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36',
];

function randomUA(): string {
  return USER_AGENTS[Math.floor(Math.random() * USER_AGENTS.length)];
}

function extractMainContent(html: string): string {
  // Try to extract from semantic containers first
  const mainSelectors = [
    /<main[^>]*>([\s\S]*?)<\/main>/gi,
    /<article[^>]*>([\s\S]*?)<\/article>/gi,
    /<div[^>]*role=["']main["'][^>]*>([\s\S]*?)<\/div>/gi,
    /<div[^>]*class=["'][^"']*content[^"']*["'][^>]*>([\s\S]*?)<\/div>/gi,
  ];

  for (const selector of mainSelectors) {
    const match = selector.exec(html);
    if (match?.[1] && match[1].length > 200) {
      return stripHtmlToText(match[1]);
    }
  }

  // Fallback: strip everything but remove nav/header/footer first
  const cleaned = html
    .replace(/<nav[\s\S]*?<\/nav>/gi, '')
    .replace(/<header[\s\S]*?<\/header>/gi, '')
    .replace(/<footer[\s\S]*?<\/footer>/gi, '')
    .replace(/<aside[\s\S]*?<\/aside>/gi, '')
    .replace(/<noscript[\s\S]*?<\/noscript>/gi, '');

  return stripHtmlToText(cleaned);
}

function stripHtmlToText(html: string): string {
  return html
    .replace(/<script[\s\S]*?<\/script>/gi, '')
    .replace(/<style[\s\S]*?<\/style>/gi, '')
    .replace(/<svg[\s\S]*?<\/svg>/gi, '')
    .replace(/<[^>]+>/g, ' ')
    .replace(/&nbsp;/gi, ' ')
    .replace(/&amp;/gi, '&')
    .replace(/&lt;/gi, '<')
    .replace(/&gt;/gi, '>')
    .replace(/&quot;/gi, '"')
    .replace(/&#39;/gi, "'")
    .replace(/\s+/g, ' ')
    .trim();
}

async function smartDirectFetch(url: string): Promise<{ markdown: string; title: string } | null> {
  try {
    const { assertSafeUrl } = await import('@/lib/guardrails');
    await assertSafeUrl(url);
  } catch {
    return null;
  }

  try {
    const res = await fetch(url, {
      headers: {
        'User-Agent': randomUA(),
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
        'Accept-Language': 'en-US,en;q=0.9',
        'Accept-Encoding': 'gzip, deflate, br',
        'Cache-Control': 'no-cache',
        'Sec-Fetch-Dest': 'document',
        'Sec-Fetch-Mode': 'navigate',
        'Sec-Fetch-Site': 'none',
        'Sec-Fetch-User': '?1',
        'Upgrade-Insecure-Requests': '1',
      },
      redirect: 'follow',
      signal: AbortSignal.timeout(15_000),
    });

    if (!res.ok) return null;

    const html = await res.text();
    const text = extractMainContent(html).slice(0, 6000);

    const titleMatch = html.match(/<title[^>]*>(.*?)<\/title>/i);
    const metaDesc = html.match(/<meta[^>]*name=["']description["'][^>]*content=["']([^"']+)["']/i);
    const pageTitle = titleMatch?.[1]?.trim() || url;

    if (text.length < 50) return null;

    const markdown = metaDesc?.[1]
      ? `${metaDesc[1].trim()}\n\n${text}`
      : text;

    return { markdown, title: pageTitle };
  } catch {
    return null;
  }
}

async function scrapeBasic(url: string): Promise<ToolResult<ScrapedPage>> {
  const policy = getPolicyForDomain(url);

  if (policy.useBrowserFallback && isPlaywrightScrapeEnabled()) {
    const pw = await scrapeWithPlaywright(url);
    if (pw) {
      const page: ScrapedPage = {
        url,
        title: pw.title,
        markdown: pw.markdown,
        excerpt: pw.markdown.slice(0, 500),
      };
      return buildToolResult<ScrapedPage>({
        data: page,
        status: 'degraded',
        source: 'Playwright (no Firecrawl key)',
        sourceUrl: url,
      });
    }
  }

  // Try Scrape.do first (if token available), then smart direct fetch
  const scrapeDo = await scrapeDoFetch(url);
  if (scrapeDo) {
    const page: ScrapedPage = { url, title: scrapeDo.title, markdown: scrapeDo.markdown, excerpt: scrapeDo.markdown.slice(0, 500) };
    return buildToolResult<ScrapedPage>({ data: page, status: 'degraded', source: 'Scrape.do (no Firecrawl key)', sourceUrl: url });
  }

  const direct = await smartDirectFetch(url);
  if (direct) {
    const page: ScrapedPage = { url, title: direct.title, markdown: direct.markdown, excerpt: direct.markdown.slice(0, 500) };
    return buildToolResult<ScrapedPage>({ data: page, status: 'degraded', source: 'Direct Scrape (fallback)', sourceUrl: url });
  }

  const page: ScrapedPage = { url, title: url, markdown: '', excerpt: 'Could not scrape this page.' };
  return buildToolResult<ScrapedPage>({ data: page, status: 'failed', source: 'Direct Scrape (failed)', sourceUrl: url });
}

export async function scrapeCompetitorPricing(productUrl: string): Promise<ToolResult<ScrapedPage>> {
  // Append /pricing if it looks like a root domain
  const pricingUrl = productUrl.replace(/\/$/, '') + '/pricing';
  return scrapePage(pricingUrl);
}

File: backend/app/tools/firecrawl.py
"""Firecrawl + scrape fallbacks — mirrors lib/tools/firecrawl.ts (core paths)."""

from __future__ import annotations

import asyncio
import random
import re
from typing import Any
from urllib.parse import urlencode

import httpx

from app.config import get_settings
from app.supabase_client import get_cached, set_cache

from .fallback import build_tool_result

BASE_URL = "https://api.firecrawl.dev/v1"
SCRAPE_DO_BASE = "https://api.scrape.do"
TIMEOUT = 15.0
SCRAPE_DO_TIMEOUT = 25.0

EXTRACT_PROMPTS: dict[str, str] = {
    "pricing": (
        "Extract pricing tiers, features per tier, costs, and billing periods. "
        "Highlight any discounts or special offers."
    ),
    "features": (
        "Extract product features, capabilities, technical details, and key differentiators."
    ),
    "competitor": (
        "Extract positioning, USPs, target audience, and competitive claims."
    ),
    "review": "Extract review ratings, sentiment, pros/cons, and buyer feedback.",
    "generic": (
        "Extract key product information, pricing, features, and any competitive claims."
    ),
}

USER_AGENTS = [
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_5) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.5 Safari/605.1.15",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:127.0) Gecko/20100101 Firefox/127.0",
    "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 14_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
]

SPA_PATTERNS = [
    re.compile(r"facebook\.com", re.I),
    re.compile(r"linkedin\.com", re.I),
    re.compile(r"twitter\.com", re.I),
    re.compile(r"x\.com", re.I),
    re.compile(r"app\.", re.I),
    re.compile(r"dashboard\.", re.I),
    re.compile(r"portal\.", re.I),
]


def _select_extract_prompt(url: str) -> str:
    lower = url.lower()
    if re.search(r"pricing|plan|billing|cost", lower):
        return EXTRACT_PROMPTS["pricing"]
    if re.search(r"feature|capability|docs|technical", lower):
        return EXTRACT_PROMPTS["features"]
    if re.search(r"review|g2|capterra|comparison", lower):
        return EXTRACT_PROMPTS["review"]
    if re.search(r"competitor|vs|alternative", lower):
        return EXTRACT_PROMPTS["competitor"]
    return EXTRACT_PROMPTS["generic"]


def _needs_js_render(url: str) -> bool:
    return any(p.search(url) for p in SPA_PATTERNS)


def _strip_html_to_text(html: str) -> str:
    text = re.sub(r"<script[\s\S]*?</script>", " ", html, flags=re.I)
    text = re.sub(r"<style[\s\S]*?</style>", " ", text, flags=re.I)
    text = re.sub(r"<svg[\s\S]*?</svg>", " ", text, flags=re.I)
    text = re.sub(r"<[^>]+>", " ", text)
    text = text.replace("&nbsp;", " ").replace("&amp;", "&")
    text = text.replace("&lt;", "<").replace("&gt;", ">")
    text = text.replace("&quot;", '"').replace("&#39;", "'")
    text = re.sub(r"\s+", " ", text).strip()
    return text


def _extract_main_content(html: str) -> str:
    for pattern in (
        r"<main[^>]*>([\s\S]*?)</main>",
        r"<article[^>]*>([\s\S]*?)</article>",
        r'<div[^>]*role=["\']main["\'][^>]*>([\s\S]*?)</div>',
        r'<div[^>]*class=["\'][^"\']*content[^"\']*["\'][^>]*>([\s\S]*?)</div>',
    ):
        match = re.search(pattern, html, flags=re.I)
        if match and match.group(1) and len(match.group(1)) > 200:
            return _strip_html_to_text(match.group(1))

    cleaned = html
    for tag in ("nav", "header", "footer", "aside", "noscript"):
        cleaned = re.sub(rf"<{tag}[\s\S]*?</{tag}>", " ", cleaned, flags=re.I)
    return _strip_html_to_text(cleaned)


def _assess_scrape_quality(markdown: str, url: str) -> dict[str, Any]:
    text = (markdown or "").strip()
    lower = text.lower()
    block_signals = (
        "access denied",
        "captcha",
        "please enable javascript",
        "sign in to continue",
        "403 forbidden",
    )
    is_block = any(sig in lower for sig in block_signals)
    is_valid = len(text) >= 100 and not is_block
    quality_score = min(1.0, len(text) / 2000) if is_valid else 0.2
    return {
        "isBlockPage": is_block,
        "isValid": is_valid,
        "qualityScore": quality_score,
    }


async def _firecrawl_fetch(
    url: str,
    extract_prompt: str,
    api_key: str,
    *,
    is_strict: bool = False,
) -> dict[str, str] | None:
    payload: dict[str, Any] = {
        "url": url,
        "formats": ["markdown", "extract"],
        "extract": {"prompt": extract_prompt},
    }
    if is_strict:
        payload.update(
            {
                "waitForSelector": "body",
                "onlyMainContent": True,
                "removeTags": ["script", "style"],
            }
        )

    try:
        async with httpx.AsyncClient(timeout=TIMEOUT) as client:
            res = await client.post(
                f"{BASE_URL}/scrape",
                headers={
                    "Content-Type": "application/json",
                    "Authorization": f"Bearer {api_key}",
                },
                json=payload,
            )
        if not res.is_success:
            return None
        raw = res.json()
        markdown = (raw.get("data") or {}).get("markdown") or ""
        title = (raw.get("data") or {}).get("metadata", {}).get("title") or url
        if len(markdown.strip()) <= 50:
            return None
        return {"markdown": markdown, "title": title}
    except Exception:
        return None


async def _scrape_do_fetch(url: str) -> dict[str, str] | None:
    settings = get_settings()
    if not settings.scrape_do_token:
        return None

    params: dict[str, str] = {
        "token": settings.scrape_do_token,
        "url": url,
        "output": "markdown",
    }
    if _needs_js_render(url):
        params["render"] = "true"
        params["waitUntil"] = "domcontentloaded"

    try:
        async with httpx.AsyncClient(timeout=SCRAPE_DO_TIMEOUT) as client:
            res = await client.get(f"{SCRAPE_DO_BASE}/?{urlencode(params)}")
        if not res.is_success:
            return None
        markdown = res.text.strip()
        if len(markdown) < 50:
            return None
        title_match = re.search(r"^#\s+(.+)$", markdown, re.M)
        title = title_match.group(1).strip() if title_match else url
        return {"markdown": markdown[:8000], "title": title}
    except Exception:
        return None


async def _smart_direct_fetch(url: str) -> dict[str, str] | None:
    try:
        async with httpx.AsyncClient(timeout=TIMEOUT, follow_redirects=True) as client:
            res = await client.get(
                url,
                headers={
                    "User-Agent": random.choice(USER_AGENTS),
                    "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                    "Accept-Language": "en-US,en;q=0.9",
                    "Cache-Control": "no-cache",
                },
            )
        if not res.is_success:
            return None
        html = res.text
        text = _extract_main_content(html)[:6000]
        title_match = re.search(r"<title[^>]*>(.*?)</title>", html, re.I | re.S)
        meta_match = re.search(
            r'<meta[^>]*name=["\']description["\'][^>]*content=["\']([^"\']+)["\']',
            html,
            re.I,
        )
        page_title = title_match.group(1).strip() if title_match else url
        if len(text) < 50:
            return None
        markdown = f"{meta_match.group(1).strip()}\n\n{text}" if meta_match else text
        return {"markdown": markdown, "title": page_title}
    except Exception:
        return None


async def _scrape_basic(url: str) -> dict[str, Any]:
    scrape_do = await _scrape_do_fetch(url)
    if scrape_do:
        page = {
            "url": url,
            "title": scrape_do["title"],
            "markdown": scrape_do["markdown"],
            "excerpt": scrape_do["markdown"][:500],
        }
        return build_tool_result(
            data=page,
            status="degraded",
            source="Scrape.do (no Firecrawl key)",
            source_url=url,
        )

    direct = await _smart_direct_fetch(url)
    if direct:
        page = {
            "url": url,
            "title": direct["title"],
            "markdown": direct["markdown"],
            "excerpt": direct["markdown"][:500],
        }
        return build_tool_result(
            data=page,
            status="degraded",
            source="Direct Scrape (fallback)",
            source_url=url,
        )

    page = {
        "url": url,
        "title": url,
        "markdown": "",
        "excerpt": "Could not scrape this page.",
    }
    return build_tool_result(
        data=page,
        status="failed",
        source="Direct Scrape (failed)",
        source_url=url,
    )


async def scrape_page(url: str) -> dict[str, Any]:
    cache_key = f"scrape:{url}"
    cached = await get_cached("firecrawl", cache_key)
    if cached:
        return {**cached, "cached": True}

    settings = get_settings()
    if not settings.firecrawl_api_key:
        return await _scrape_basic(url)

    extract_prompt = _select_extract_prompt(url)
    attempts_made = 0
    result_data: dict[str, str] | None = None

    result_data = await _firecrawl_fetch(url, extract_prompt, settings.firecrawl_api_key)
    attempts_made += 1

    if not result_data:
        await asyncio.sleep(1.0)
        result_data = await _firecrawl_fetch(
            url, extract_prompt, settings.firecrawl_api_key, is_strict=True
        )
        attempts_made += 1

    if not result_data:
        await asyncio.sleep(1.5)
        result_data = await _scrape_do_fetch(url)
        attempts_made += 1

    if not result_data:
        await asyncio.sleep(2.0)
        result_data = await _smart_direct_fetch(url)
        attempts_made += 1

    markdown = result_data["markdown"] if result_data else ""
    title = result_data["title"] if result_data else url

    quality = _assess_scrape_quality(markdown, url)
    page = {
        "url": url,
        "title": title,
        "markdown": markdown,
        "excerpt": markdown[:500],
    }

    status: str = "ok"
    confidence = 0.85

    if not markdown.strip():
        status = "failed"
        confidence = 0.15
    elif quality["isBlockPage"]:
        status = "degraded"
        confidence = 0.35
    elif not quality["isValid"]:
        status = "degraded"
        confidence = quality["qualityScore"] * 0.7
    elif attempts_made > 1:
        status = "degraded"
        confidence = 0.7

    result = build_tool_result(
        data=page,
        status=status,  # type: ignore[arg-type]
        source="Firecrawl",
        source_url=url,
        confidence_override=confidence,
    )
    await set_cache("firecrawl", cache_key, result)
    return result


async def scrape_competitor_pricing(product_url: str) -> dict[str, Any]:
    pricing_url = product_url.rstrip("/") + "/pricing"
    return await scrape_page(pricing_url)
